### ISSUE TITLE
Some general suggestions about making railcraft a more viable option for remote logistics in survival gaming

### ISSUE THREAD

**Issue by wormzjl** [Railcraft/Railcraft#1508]
Hello, I'm a hardcore modded MC player and modpack maker. I've been using railcraft for years and always liked the idea of using trains for logistics on a larger scale. Last year I've released a modpack which enforces such activity by requiring players to setup remote mining bases and use rails for ore/oil transportation. The result, however, was not very satisfying, as players tried to evade using rails as much as they can, even by carrying resources manually over large distances.
After some observation, I've concluded that there are 3 points that needs to be considered to make railcraft more viable as a remote logistics option in survival mode for modern modded minecraft.
1. Accessibility
2. Capability/efficiency
3. Performance

Accessibility:
The most important problem with railcraft rails is its accessibility. With a large selection of rail parts, it's very easy for new players to get lost. From my point of view, it could be helpful to have an in-game documentation with example builds, emphasizing some of the most useful rails. Players can learn about the function of other rail parts later if needed.
Apart from documentations, the biggest reason that people don't use rails is the amount of effort required to establish such systems. You need to first clear the path, build bridges when necessary, laying down the tracks, place and configure signals and related systems when necessary, etc... If not using locomotives (such as HS tracks), you have to place booster tracks and redstone torches, which could be a hell of work...The tunnel bore machine could ease the pain slightly, but still it has many limitations, such as cannot place rails across oceans due to the max vertical distance for gravel placement (and it's consumption), can be blocked by water/lava undergrounds, can only place simple rails, etc.… Using 3rd party tools such as BC builder could help, but this adds extra dependency and knowledge gate, which won't be helpful improving the current situation. In my opinion, this need to be dealt with in railcraft’s framework.
What I'm suggesting, is either improve the current TBM design, or create a new "advanced TBM" for easier rail-laying tasks. I have a couple of rough ideas:
-	The ability to cross large chasms/oceans on its own without the need for constant player surveillance. This requires the TBM to be able to place solid blocks instead of only gravels. I do like the idea of realism in my own builds, but IMO it could be better to be a config option instead of forced upon.
-	Don’t get jammed by water/lava. Give it the ability to place/replace wall blocks and dig through such fluid lakes if possible.
-	The ability to place rails and other blocks (RS torches, light sources, signals…) on pattern, like using booster track instead of normal tracks every X blocks.
-	Double lane laying, with current TBM it’s very time consuming.
-	Easy turning to left or right instead of needing to break and replace/reconfigure the whole TBM.
The objective is to make rail-laying fast and automatable for the sake of survival gaming.
There is also a minor problem for RC which is microcrafting, having complex recipes for rails and signal posts will prevent some players from looking inside the mod. But this issue can be dealt with on pack maker’s side, so it’s just something to keep in mind.

Capability/efficiency:
When using locomotive for transportation, the speed is always an issue. I’d suggest having some sort of high-speed locomotive since steam locos doesn’t work well with HS tracks last time I checked. 
A config option for HS track explosion will be welcome, it’s extremely difficult to make a rail line free of mobs in survival, and there are various things that might occur which can break it. When it breaks, finding the explosion point is also painful.
Also, diesel locomotive anyone?

Performance:
When using locomotives for transportation, anchor carts are not doubt essential, but rapidly loading/unloading chunks could cause performance issues, especially on servers with multiple lines running. Instead of loading 3x3 chunks around the cart, maybe there is a way to only load chunks on its path?
Furthermore, signals breaks between loaded/unloaded chunks, I wonder if there is any method to bypass such limitations.
The electric locomotive, while one doesn’t need to fuel it, introduces more problem than it solves. One of my major concern is performance at larger scale. Having a MFSU cart instead of electric rails ticking all the time is surely more performance-friendly. Once again, one need to make compromises between realism and gameplay, a config option about self-powered electric locomotive slowdown rate would be great.

TL;DR : In-game documentation, fully automatable rail works, high-speed improvements, chunk loading tweaks.
So that’s all I have in mind for now, we can discuss in more details if there is anything appears interesting to you.

**Comment by wormzjl** [comment]
Also, https://github.com/Railcraft/Railcraft/issues/1460 is a blocking issue if you really want to use RC for unmanned transportation.

**Comment by liach** [comment]
Great issue for discussion.

I added some advancements. Dunno if it helps for documentation.

The charge system's tick is not that costly compared to carts. A few networks are ticked to update instead of all positions (as I recall).

**Comment by CovertJaguar** [comment]
Yes, the Charge system was specifically redesigned a while back so that basically only generators need to tick. The blocks that just transmit or store charge are normal non-ticking blocks, they don't even have tile entities.

The Worldspike carts actually currently load 5x5 areas because 3x3 wasn't large enough to make junctions work correctly. There might be an argument here for separating the Signal Network from the world like the Charge Network separates nodes from the world. I don't think it would be a huge change, the logic just needs to move from the tile to a collection of objects that the tile can look up and update. Their state should be stored like Batteries are as well. This would probably improve overall reliability of the signals.

Locomotives work fine with HS Rails, you just need to design the HS tracks the same way you would as if you didn't have a locomotive. The Locomotive by itself isn't capable of switching the train into HighSpeed mode, its the transition tracks that do that. Increasing the speed of locomotives is problematic due to engine limitations. The Reinforced track already has to do a few hacks to keep carts from falling off corners and missing slopes. This is why HS track has to travel in a straight line.
